v2/internal/testcommon/samples_tester.go

```diff
@@ -0,0 +1,226 @@
+		tc:            tc,
+		decoder:       serializer.NewCodecFactory(tc.GetScheme()).UniversalDecoder(),
+		UseRandomName: useRandomName,
+	}
+}
+
+func (t SamplesTester) LoadSamples(rg *resources.ResourceGroup, group string) (*SampleObject, *SampleObject) {
+
+	samples := NewSampleObject()
+	depSamples := NewSampleObject()
+
+	groupPath := getGroupPath(group)
+
+	err := filepath.Walk(groupPath,
+		func(filePath string, info os.FileInfo, err error) error {
+
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() && !isExclusion(filePath) {
+				//if !info.IsDir() && !isExclusion(filePath) {
+
+				sample := t.getObjectFromFile(filePath)
+
+				sample.SetNamespace(t.tc.Namespace)
+
+				if !strings.Contains(filePath, RefsPackage) {
+					if t.UseRandomName {
+						sample.SetName(t.tc.NoSpaceNamer.GenerateName(""))
+					}
+					handleObject(sample, rg, &samples)
+				} else {
+					// We don't change name for dependencies as they have fixed refs which we can't access inside the object
+					// need to make sure we have right names for the refs in samples
+					handleObject(sample, rg, &depSamples)
+				}
+			}
+
+			return nil
+		})
+
+	t.tc.Expect(err).To(BeNil())
+
+	// We process the child objects once we have all the parent objects in the tree already
+	handleChildObjects(t.tc, &samples)
+	handleChildObjects(t.tc, &depSamples)
+
+	return &samples, &depSamples
+}
+
+func (t SamplesTester) getObjectFromFile(path string) genruntime.ARMMetaObject {
+	jsonMap := make(map[string]interface{})
+
+	byteData, err := ioutil.ReadFile(path)
+	t.tc.Expect(err).To(BeNil())
+
+	jsonBytes, err := yaml.ToJSON(byteData)
+	t.tc.Expect(err).To(BeNil())
+
+	err = json.Unmarshal(jsonBytes, &jsonMap)
+	t.tc.Expect(err).To(BeNil())
+
+	// We need unstructured object here to fetch the correct GVK for the object
+	unstructuredObj := unstructured.Unstructured{Object: jsonMap}
+
+	obj, err := genruntime.NewObjectFromExemplar(&unstructuredObj, t.tc.GetScheme())
+	t.tc.Expect(err).To(BeNil())
+
+	byteData, err = json.Marshal(unstructuredObj.Object)
+	t.tc.Expect(err).To(BeNil())
+
+	err = runtime.DecodeInto(t.decoder, byteData, obj)
+
+	return obj.(genruntime.ARMMetaObject)
+}
+
+func handleObject(sample genruntime.ARMMetaObject, rg *resources.ResourceGroup, samples *SampleObject) {
+	if sample.Owner().Kind == resolver.ResourceGroupKind {
+		handleParentObject(sample, rg, samples.SamplesMap)
+	} else {
+		samples.childObjects[sample.GetObjectKind().GroupVersionKind().String()] = sample
+	}
+}
+
+func handleChildObjects(tc *KubePerTestContext, samples *SampleObject) {
+	for gvk, sample := range samples.childObjects {
+
+		specField := reflect.ValueOf(sample.GetSpec())
+		originalVersion := specField.MethodByName("OriginalVersion").Call([]reflect.Value{})
+
+		tc.Expect(originalVersion[0]).ToNot(BeNil())
+		tc.Expect(originalVersion).ToNot(HaveLen(0))
+		tc.Expect(originalVersion[0].String()).ToNot(BeEmpty())
+
+		version := originalVersion[0].String()
```

Comment:
No need to monkey around with reflection; all our code generated resources meet the `genruntime.GroupVersionKindAware` interface by implementing `OriginalGVK()`, like this:

``` go
// OriginalGVK returns a GroupValueKind for the original API version used to create the resource
func (machine *VirtualMachine) OriginalGVK() *schema.GroupVersionKind {
	return &schema.GroupVersionKind{
		Group:   GroupVersion.Group,
		Version: machine.Spec.OriginalVersion(),
		Kind:    "VirtualMachine",
	}
}
```

So you should be able to do a simple typecast to `genruntime.GroupVersionKindAware` and call it directly.

Reply:
Using the interface is also safer - not every resource has `OriginalVersion` as a method; some have a field with the same name.